### DISCUSSION TITLE
udev?

### DISCUSSION THREAD

**Discussion by desultory** [anatol/booster#269] [General]
Hi, I'm the author of this: https://github.com/desultory/ugrd
I'm wondering how booster handles udev stuff, I'd like to do similar to make systemd happy.

Is there any clear spec of what booster does under /run to make udev aware of the cryptsetup mounts it makes?

**Comment by anatol** [comment]
Hello

`booster` does not use udev library. Instead it sets up a listener for uevents (which uses netlink protocol). And then processes uevents that are interesting to booster (such as a drive discovery event).

The library that performs uevent listening/parsing is here https://github.com/anatol/go-udev

**Reply by desultory** [reply]
Do you know if there is a spec that lists exactly what needs to be done? I'm considering writing some kind of stub that simply makes a few files in /run/udev to signal to systemd that the luks volume is online.

**Reply by desultory** [reply]
reading that project I don't see how it adds "db" entries in /run, I think this is all done in booster. how did you determine what needed to be done here?

**Reply by anatol** [reply]
> Do you know if there is a spec that lists exactly what needs to be done?

Could you please provide a more specific question what are you looking for?

If udev code then here is the booster part https://github.com/anatol/booster/blob/master/init/udev.go#L110 Booster does not use `udev` library itself and listens for udev events directly via netlink.

For the LUKS unlocking logic - it mostly concentrated in https://github.com/anatol/luks.go and booster calls that API.

**Reply by anatol** [reply]
> reading that project I don't see how it adds "db" entries in /run, I think this is all done in booster. how did you determine what needed to be done here?

Here is the code that adds a db entry to `/run` https://github.com/anatol/booster/blob/master/init/udev.go#L244

**Reply by desultory** [reply]
> > Do you know if there is a spec that lists exactly what needs to be done?
> 
> Could you please provide a more specific question what are you looking for?
> 
> If udev code then here is the booster part https://github.com/anatol/booster/blob/master/init/udev.go#L110 Booster does not use `udev` library itself and listens for udev events directly via netlink.
> 
> For the LUKS unlocking logic - it mostly concentrated in https://github.com/anatol/luks.go and booster calls that API.

Thanks for the info. I think I have this mostly sorted out. I've been trying to figure this out for months. My only concern is that LVM/redhat devs seem to say that doing something like this isn't very robust. TBH I think it's an issue with the lvm udev rules being lacking. https://github.com/lvmteam/lvm2/issues/153

It would be nice if this was not required. And FWIW this issue applies to LUKS alone as well. This issue manifest any time you mount some DM type device in the initrd without _real_ udev, and then boot into a modern systemd machine. Other inits don't need this extra help, and the "solution" is to pull the entirety of udev into the initrd.

**Comment by desultory** [comment]
I made this: https://github.com/desultory/ugrd/blob/main/src/ugrd/fs/fakeudev.py which is based on: https://github.com/anatol/booster/blob/master/init/udev.go#L251 and seems to function
